VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/Strings/StringPlaceholdersInWrongOrder/StringPlaceholdersInWrongOrderAnalyzer.cs

```diff
@@ -0,0 +1,116 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace VSDiagnostics.Diagnostics.Strings.StringPlaceholdersInWrongOrder
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class StringPlaceholdersInWrongOrderAnalyzer : DiagnosticAnalyzer
+    {
+        private static readonly string Category = VSDiagnosticsResources.StringsCategory;
+        private const string DiagnosticId = nameof(StringPlaceholdersInWrongOrderAnalyzer);
+        private static readonly string Message = VSDiagnosticsResources.StringPlaceholdersInWrongOrderMessage;
+        private const DiagnosticSeverity Severity = DiagnosticSeverity.Warning;
+        private static readonly string Title = VSDiagnosticsResources.StringPlaceholdersInWrongOrderTitle;
+
+        internal static DiagnosticDescriptor Rule => new DiagnosticDescriptor(DiagnosticId, Title, Message, Category, Severity, true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.InvocationExpression);
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var invocation = context.Node as InvocationExpressionSyntax;
+            if (invocation == null)
+            {
+                return;
+            }
+
+            // Verify we're dealing with a string.Format() call
+            var memberAccessExpression = invocation.Expression as MemberAccessExpressionSyntax;
+            if (memberAccessExpression != null)
+            {
+                var invokedType = context.SemanticModel.GetSymbolInfo(memberAccessExpression.Expression);
+                var invokedMethod = context.SemanticModel.GetSymbolInfo(memberAccessExpression.Name);
+                if (invokedType.Symbol == null || invokedMethod.Symbol == null)
+                {
+                    return;
+                }
+
+                if (invokedType.Symbol.MetadataName != typeof(string).Name || invokedMethod.Symbol.MetadataName != nameof(string.Format))
+                {
+                    return;
+                }
+            }
+
+            // Verify the format is a literal expression and not a method invocation or an identifier
+            // The overloads are in the form string.Format(string, object[]) or string.Format(CultureInfo, string, object[])
+            if (invocation.ArgumentList == null || invocation.ArgumentList.Arguments.Count < 2)
+            {
+                return;
+            }
+
+            var firstArgument = invocation.ArgumentList.Arguments[0];
+            var secondArgument = invocation.ArgumentList.Arguments[1];
+
+            var firstArgumentSymbol = context.SemanticModel.GetSymbolInfo(firstArgument.Expression);
```

Comment:
Ahhh it just clicked. When it looks up the method symbol, method resolution will already have happened so it will already have the correct method overload. Seems like this should do it just fine then:

```csharp
Debug.Assert(invocation.Expression is MemberAccessExpressionSyntax);
var memberAccessExpression = (MemberAccessExpressionSyntax) invocation.Expression;
var invokedMethod = context.SemanticModel.GetSymbolInfo(memberAccessExpression.Name);
var methodSymbol = invokedMethod.Symbol as IMethodSymbol;

var formatParam = methodSymbol?.Parameters.FirstOrDefault(x => x.Name == "format");
if (formatParam == null)
{
	return;
}

var formatIndex = formatParam.Ordinal;
```